Signup Customisation [RFC] (#1395)
wasp-lang/wasp: web/docs/auth/overview.md

</TabItem>
</Tabs>

## Customizing the Signup Process

> Sometimes you want to include **extra fields** in your signup process, like first name and last name.

Author: We have dropped hints that they can do that:
- here https://wasp-lang.dev/docs/auth/email#fields-in-the-email-dict
- and here https://wasp-lang.dev/docs/auth/username-and-pass#customizing-the-auth-flow

I could add an extra note here as well 👍

> Sometimes you want to include **extra fields** in your signup process, like first name and last name.

Reviewer: Ok, whatever you think makes sense!

In Wasp, in this case:

- You need to define the new backend fields

> - You need to customize the `SignupForm`

Author: Adjusted the language 👍

> Other times, you might need to add some **extra UI** elements to the form, like a checkbox for terms of service. In this case, customizing only the UI components is enough.

Author: > Avoid language that invalidate struggle, such as "easy", "just", "obviously", etc. For reference, see [Words To Avoid in Educational Writing](https://css-tricks.com/words-avoid-educational-writing/).

https://github.com/vuejs/v2.vuejs.org/blob/master/writing-guide.md

> Other times, you might need to add some **extra UI** elements to the form, like a checkbox for terms of service. In this case, customizing only the UI components is enough.

Reviewer: Sure, that makes sense, but this wasn't for the purpose of making it sound easy, the purpose was to make a contrast compared to the previous part where we had to do two things, while here we do only one thing. When I was reading it the first time, I thought that is not very clear, that we are doing only half of the job here.
I get what they mean with not using "just", but I don't think this case falls under that. Certainly not all usages of "just" should be eliminated, just those that invalidate struggle, like "now let's just implement this thing" or "the only thing left is just running the command".
Maybe there is another way to indicate this is only part of what was needed above it?

> Other times, you might need to add some **extra UI** elements to the form, like a checkbox for terms of service. In this case, customizing only the UI components is enough.

Author: My bad, I jumped right at the word without understanding that real meaning behind the change. The change makes a lot of sense 👍

Let's see how to do both.

### 1. Defining Extra Backend Fields

> If we want to **save** some extra fields in our signup process, we need to tell our backend they exist.

Author: Sure thing, I can adjust the language, although for me the two are interchangeable. Like client and frontend.

> If we want to **save** some extra fields in our signup process, we need to tell our backend they exist.

Reviewer: Right, but we do use client and server mostly. The reason why is that client and server are more precise terms. Frontend and backend can be quite relative and mean many things: backend of compiler for example.

We do that by defining an object where the keys represent the field name, and the values are functions that receive the data sent from the client and return the value of the field.

> First, we add the `auth.signup.additionalFields` field in our `main.wasp` file:

Reviewer: ```suggestion
First, we add the `auth.signup.additionalFields` field in our `main.wasp` file, via which we provide the path to the JS/TS file where we are going to define these extra fields.
```

Reviewer: You don't think it is good? The reason I put this is because I found it unclear as to which value that field should have. You set it to `"@server/auth/signup.js"`, but why, what is the intention? Does it have to that file name, can it be something else?
If you feel it is clear, ok, leave it as it is.

Author: Based on the "Show, don't tell" principle, this sentence feels redundant. It's clear from the context that we are putting a JS/TS file there and it's clear that in that file we are defining the fields.

Reviewer: To me that wasn't clear from the context, not until I read further along. But ok, I will leave it to you.

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```wasp title="main.wasp" {9-11}
app crudTesting {
  // ...
  auth: {
    userEntity: User,
    methods: {
      usernameAndPassword: {},
    },
    onAuthFailedRedirectTo: "/login",
    signup: {
      additionalFields: import { fields } from "@server/auth/signup.js",
    },
  },
}

entity User {=psl
  id Int @id @default(autoincrement())
  username String @unique
  password String
  address String?
psl=}
```

> Then we'll export the `fields` object from the `server/auth/signup.js` file:

Reviewer: ```suggestion
Then we'll define and export the `fields` object from the `server/auth/signup.js` file:
```

  // ...
  auth: {
    userEntity: User,
    methods: {
      usernameAndPassword: {},
    },
    onAuthFailedRedirectTo: "/login",
    signup: {
      additionalFields: import { fields } from "@server/auth/signup.js",
    },
  },
}

entity User {=psl
  id Int @id @default(autoincrement())
  username String @unique
  password String
  address String?
psl=}
```

Then we'll export the `fields` object from the `server/auth/signup.ts` file:

```ts title="server/auth/signup.ts"
import { defineAdditionalSignupFields } from "@wasp/auth/index.js";

export const fields = defineAdditionalSignupFields({
  address: async (data) => {
    const address = data.address;
    if (typeof address !== "string") {
      throw new Error("Address is required");
    }
    if (address.length < 5) {
      throw new Error("Address must be at least 5 characters long");
    }
    return address;
  },
});
```

</TabItem>
</Tabs>

<small>

Read more about the `fields` object in the [API Reference](#signup-fields-customization).
</small>

Keep in mind, that these field names need to exist on the `userEntity` you defined in your `main.wasp` file e.g. `address` needs to be a field on the `User` entity. 

> The field function will receive the data sent from the client and it needs to return the value of the field. If the field is invalid, the function should throw an error.

Reviewer: ```suggestion
The field function will receive the data sent from the client and it needs to return the value that will be saved into the database. If the field is invalid, the function should throw an error.
```

      .string({
        required_error: 'Address is required',
        invalid_type_error: 'Address must be a string',
      })
      .min(10, 'Address must be at least 10 characters long');
    const result = AddressSchema.safeParse(data.address);
    if (result.success === false) {
      throw new Error(result.error.issues[0].message);
    }
    return result.data;
  },
})
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="server/auth/signup.ts"
import { defineAdditionalSignupFields } from '@wasp/auth/index.js'
import * as z from 'zod'

export const fields = defineAdditionalSignupFields({
  address: (data) => {
    const AddressSchema = z
      .string({
        required_error: 'Address is required',
        invalid_type_error: 'Address must be a string',
      })
      .min(10, 'Address must be at least 10 characters long');
    const result = AddressSchema.safeParse(data.address);
    if (result.success === false) {
      throw new Error(result.error.issues[0].message);
    }
    return result.data;
  },
})
```
</TabItem>
</Tabs>
</details>

:::

Now that we defined the fields, Wasp knows how to:

1. Validate the data sent from the client

> 2. Save the data to the database

Reviewer: Nice, this is clear. I would periods at the end of sentences though. Remember, if sentence starts with capital letter, it should end with period (or something similar).

Author: Adjusted this piece as well 👍

  address: async (data) => {
    const address = data.address;
    if (typeof address !== "string") {
      throw new Error("Address is required");
    }
    if (address.length < 5) {
      throw new Error("Address must be at least 5 characters long");
    }
    return address;
  },
});
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp title="main.wasp" {9-11}
app crudTesting {
  // ...
  auth: {
    userEntity: User,
    methods: {
      usernameAndPassword: {},
    },
    onAuthFailedRedirectTo: "/login",
    signup: {
      additionalFields: import { fields } from "@server/auth/signup.js",
    },
  },
}
```

Then we'll export the `fields` object from the `server/auth/signup.ts` file:

```ts title="server/auth/signup.ts"
import { defineAdditionalSignupFields } from "@wasp/auth/index.js";

export const fields = defineAdditionalSignupFields({
  address: async (data) => {
    const address = data.address;
    if (typeof address !== "string") {
      throw new Error("Address is required");
    }
    if (address.length < 5) {
      throw new Error("Address must be at least 5 characters long");
    }
    return address;
  },
});
```

</TabItem>
</Tabs>

> The `fields` object is an object where the keys represent the field name, and the values are functions which receives the data sent from the client and return the value of the field.

Reviewer: ```suggestion
The `fields` object is an object where the keys represent the field name, and the values are functions which receive the data sent from the client and return the value of the field.
```

If the field value is invalid, the function should throw an error.

The field details object has the following properties:

> 

Reviewer: This stuff is same as above -> is there any way we could avoid this duplication?

Author: If we want to keep the API reference independent, there will be some duplication always.

Reviewer: But this is identical, right? The more duplication we have, more chances of it going out of sync. I know we can't remove all duplication, some of it is semantical, but this seems like literal copy paste, that is why I am asking, also for the future, if not now.
We could maybe, when we have duplication like this, extract it into a React component?

Author: > some of it is semantical, but this seems like literal copy paste

Once we pull out this content into a API Reference component, then it needs to be standalone. I could pull out this piece of repetition into a tiny React component. But then we need to make sure that tiny piece fits in both contexts over time (for example, it can't refer to something above or below itself - since it's used in two difference places) which think is worse than having some duplication.

Reviewer: Sure, if it can't be the same in both places, that is fine, but in case we do have pieces that are completely the same, I think it is worth it investigating extracting them.
